src/Microsoft.DotNet.Build.Tasks/PreprocessFile.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class PreprocessFile : Task
+    {
+        [Required]
+        public string SourceFile { get; set; }
+
+        [Required]
+        public string OutputFile { get; set; }
+
+        public string[] Defines { get; set; }
+
+        public override bool Execute()
+        {
+            HashSet<string> defineSet = new HashSet<string>(Defines ?? new string[] { }, StringComparer.Ordinal);
+            Stack<bool> preprossorStack = new Stack<bool>();
+
+            preprossorStack.Push(true);
+
+            using (StreamWriter outputStream = new StreamWriter(File.Open(OutputFile, FileMode.Create, FileAccess.Write, FileShare.None), Encoding.UTF8))
+            {
+                foreach (string line in File.ReadLines(SourceFile))
+                {
+                    if (line.Trim().StartsWith("#if ", StringComparison.Ordinal))
+                    {
+                        string defineName = line.Trim().Substring(3).Trim();
```

Comment:
We are doing line.Trim() a couple of times here -- maybe capture that in a local?